apps/emqx_durable_storage/src/emqx_ds_storage_skipstream_lts.erl

```diff
@@ -0,0 +1,739 @@
+        {ok, TS, DSKey, Msg0} ->
+            %% ?tp(notice, skipstream_loop_result, #{r => ok, ts => TS, key => Key}),
+            Message = enrich(Ctx, DSKey, Msg0),
+            It = It0#it{ts = TS},
+            next_loop(Ctx, It, BatchSize - 1, next, [{DSKey, Message} | Acc])
+    end.
+
+finalize_loop(It0, Op, Acc) ->
+    case Op of
+        next -> NextTS = It0#it.ts + 1;
+        {seek, NextTS} -> ok
+    end,
+    It = It0#it{ts = NextTS},
+    {ok, It, lists:reverse(Acc)}.
+
+next_step(
+    S, StaticIdx, CompressedTF, [#l{hash = Hash, handle = IH, n = N} | Iterators], ExpectedTS, Op
+) ->
+    Result =
+        case Op of
+            next ->
+                inc_counter(?DS_SKIPSTREAM_LTS_NEXT),
+                rocksdb:iterator_move(IH, next);
+            {seek, TS} ->
+                inc_counter(?DS_SKIPSTREAM_LTS_SEEK),
+                rocksdb:iterator_move(IH, {seek, mk_key(StaticIdx, N, Hash, TS)})
+        end,
+    case Result of
+        {error, invalid_iterator} ->
+            none;
+        {ok, Key, Blob} ->
+            case match_key(StaticIdx, N, Hash, Key) of
+                false ->
+                    %% This should not happen, since we set boundaries
+                    %% to the iterators, and overflow to a different
+                    %% key prefix should be caught by the previous
+                    %% clause:
+                    none;
+                NextTS when ExpectedTS =:= undefined; NextTS =:= ExpectedTS ->
+                    %% We found a key that corresponds to the
+                    %% timestamp we expect.
+                    %% ?tp(notice, ?MODULE_STRING "_step_hit", #{
+                    %%     next_ts => NextTS, expected => ExpectedTS, n => N
+                    %% }),
+                    case Iterators of
+                        [] ->
+                            %% This is data stream as well. Check
+                            %% message for hash collisions and return
+                            %% value:
+                            Msg0 = deserialize(S, Blob),
+                            case emqx_topic:match(Msg0#message.topic, CompressedTF) of
+                                true ->
+                                    inc_counter(?DS_SKIPSTREAM_LTS_HIT),
+                                    {ok, NextTS, Key, Msg0};
+                                false ->
+                                    %% Hash collision. Advance to the
+                                    %% next timestamp:
+                                    inc_counter(?DS_SKIPSTREAM_LTS_HASH_COLLISION),
+                                    {seek, NextTS + 1}
+                            end;
+                        _ ->
+                            %% This is index stream. Keep going:
+                            next_step(S, StaticIdx, CompressedTF, Iterators, NextTS, {seek, NextTS})
+                    end;
+                NextTS when NextTS > ExpectedTS, N > 0 ->
```

Comment:
Nit. The `NextTS when NextTS > ExpectedTS, N == 0` clause is notably missing. I guess we don't really expect that because it would mean our storage is corrupted and atomicity of some batch write was lost, but I think having a comment would help.

Reply:
Replication layer must not advance NextTS past the point that was replicated.

Reply:
Added a comment.